[feature/migration] Migration from the app with same Bundle ID

mneuwert opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.
-->

## Description
We need to implement a migration path for apps where we migrate from the legacy to a new app and use same bundle ID.

## Related Issue
#270 

## Motivation and Context
See description in the issue

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## QA

Test plan:

https://github.com/owncloud/QA/blob/master/Mobile/iOS-app/Version%2011.4.0/Migration%20from%20legacy.md

BUGS & IMPROVEMENTS

- [ ] (1) https://github.com/owncloud/ios-app/pull/660#issuecomment-657548788

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/ios-app?pullRequest=660) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/ios-app?pullRequest=660) before we can accept your contribution.<br/><hr/>**Michael Neuwert** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/ios-app?pullRequest=660) it.</sub>

felix-schwarz commented:
@mneuwert I accidentally pushed an incomplete local merge with the changes from `master` here, alongside the commit to `feature/open_private_link`. Sorry! I subsequently completed the merge. Please feel free to keep or remove these commits.

mneuwert commented:
@felix-schwarz I think it depends on in which milestone it shall be included... if it is not 1.4, then we can keep this branch updated to master

felix-schwarz commented on ownCloud/Migration/Migration.swift:
```diff
@@ -0,0 +1,517 @@
+								self.migrationQueue.async {
+									if let userId = rowDict["id"] as? Int,
+										let serverURL = rowDict["url"] as? String,
+										let credentials = self.getCredentialsDataItem(for: userId) {
```
For reasons unbeknownst to me (app prefix identifier, group ID and keychain entry name seem to match in the debugger), the new app trying to access the credentials from the legacy app (whose bundleID I changed to `com.owncloud.ios-app`) fails at `SecItemCopyMatching` with a `errSecItemNotFound` return value, resulting in a `nil` value being returned from `getCredentialsDataItem`.

The `if` condition above subsequently fails, leading to this blank view:
<img width="386" alt="Bildschirmfoto 2020-05-11 um 23 24 19" src="https://user-images.githubusercontent.com/639669/81613316-8b75d700-93de-11ea-9361-ec7818799eeb.png">

mneuwert replied:
@felix-schwarz I will revisit this part and fix the handling

mneuwert replied:
@felix-schwarz Fixed it, now the bookmark migration will be shown as failed in case no data can't be retrieved from keychain. The question is however: should we allow migrating accounts for which no authorisation data can't be read still? And just show a warning that user will have to re-authorize this bookmark?

felix-schwarz replied:
@mneuwert In the wild, accounts for which no auth data can be read should be relatively rare. Although, of course, they can naturally occur as a byproduct if someone restores their iPhone from an unencrypted backup (restoring the app's data, but not its keychain).

If it's not too much effort, handling this scenario as well would of course be nice. But it clearly is a niche case.

mneuwert replied:
@felix-schwarz Actually my hope to do something about such account has vanished after I double-checked the code.. Legacy app has only stored URL in the database and user ID (which is just a counter) which referenced data blob in the keychain (storing username, auth method, auth token etc). So If keychain is lost, URL IMO is not sufficient.

felix-schwarz commented:
Unfortunately, I couldn't test the actual migration due to the keychain access issue. Not sure why the SQLite DB is accessible by the new app while the keychain items can't be found.

mneuwert commented:
@felix-schwarz thanks for feedback. Apart from bundle identifiers, I had to change all the .entitlements files manually to put the old and new app into the same app-group. Worked for me so far.

mneuwert commented on ownCloud/Migration/Migration.swift:
```diff
@@ -0,0 +1,517 @@
+
+				if let issue = issue {
+					if issue.level != .error, supportedAuthMethods == nil {
+						issue.approve()
```
@felix-schwarz yes, agree.. my assumption however was that if user was using the account in the legacy app, then It can be migrated AS IS. Also probably accepting a warning about using `http` is fine, but certificate acceptance is of course more sensitive topic, since we are not comparing here if it is same certificate which was accepted in the legacy app. 

So may be using similar mechanism as used during bookmark creation utilising `ConnectionIssueViewController` could be used and every warning should be re-accepted.

mneuwert replied:
@felix-schwarz I've implemented the handling I suggested.

mneuwert commented on ownCloud/Migration/Migration.swift:
```diff
@@ -0,0 +1,517 @@
+					}
+
+				} else {
+					Log.error(tagged: ["MIGRATION"], "Couldn't open legacy database, error \(String(describing: err))")
```
@felix-schwarz good point...

mneuwert replied:
fixed

mneuwert commented on ownCloud/Migration/Migration.swift:
```diff
@@ -0,0 +1,517 @@
+//
+
+/*
+* Copyright (C) 2019, ownCloud GmbH.
```
fixed

mneuwert commented on ownCloud/Migration/LegacyCredentials.swift:
```diff
@@ -0,0 +1,109 @@
+//
+
+/*
+* Copyright (C) 2019, ownCloud GmbH.
```
fixed

mneuwert commented on ownCloud/Migration/MigrationActivityCell.swift:
```diff
@@ -0,0 +1,143 @@
+//
+
+/*
+ * Copyright (C) 2019, ownCloud GmbH.
```
fixed

mneuwert commented on ownCloud/Migration/MigrationViewController.swift:
```diff
@@ -0,0 +1,118 @@
+//
+
+/*
+ * Copyright (C) 2019, ownCloud GmbH.
```
fixed

felix-schwarz commented:
Looking good so far, but there's one major (merge) issue: previously, in that branch there were reverts of previous merges with `origin/master` (7a050bb and 5e10a3d). Since they came after the merges with `origin/master`, they not only rolled back all changes merged in from `master` (and thereby removed bugfixes and new features), but also - due to their "later" place in commit history - also prevented that another merge with `origin/master` can bring them back.

To fix that, I removed the merge and merge revert commits from the branch's history, did a clean merge with `origin/master` and then force-pushed the result to bring back all the bugfixes and new features that it contains.

Unfortunately, it looks like a `git pull` on your end resulted in https://github.com/owncloud/ios-app/pull/660/commits/e7b23d45a12a7b2d2029c20753056068e550ba93, which again brought back a version of the app's other sources that lacks recent bugfixes and new features (like iPadOS pointer support) and likely also prevents them from being merged back in:

```
Updating db0d4d14..d0f06efb
Fast-forward
 .swiftlint.yml                                                                                   |   1 -
 docs/bin/cli                                                                                     | 245 -----------------------------------------------
 docs/books/ownCloud_iOS_App_Manual.adoc                                                          |   2 +
 docs/modules/ROOT/assets/images/21_File_list.png                                                 | Bin 96756 -> 80330 bytes
 docs/modules/ROOT/assets/images/21_File_list_annotated.png                                       | Bin 100164 -> 107856 bytes
 docs/modules/ROOT/assets/images/21_File_list_parent.jpg                                          | Bin 105053 -> 0 bytes
 docs/modules/ROOT/assets/images/21_File_list_parent.png                                          | Bin 103215 -> 95546 bytes
 docs/modules/ROOT/assets/images/directory-actions.png                                            | Bin 120269 -> 52390 bytes
 docs/modules/ROOT/assets/images/file-actions.jpg                                                 | Bin 64816 -> 0 bytes
 docs/modules/ROOT/assets/images/index-bar-with-callout.png                                       | Bin 58119 -> 0 bytes
 docs/modules/ROOT/assets/images/index-bar.png                                                    | Bin 146383 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/all-available-offline-items.png                  | Bin 125482 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/available-offline-badge-closeup-with-callout.png | Bin 191801 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/available-offline-badge-closeup.png              | Bin 31355 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/available-offline-badge.png                      | Bin 163927 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/available-offline-items.png                      | Bin 55828 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/available-offline-logo.png                       | Bin 983 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/file-available-offline.png                       | Bin 31257 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/folder-action.png                                | Bin 155874 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/make-available-offline.png                       | Bin 32107 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/make-unavailable-offline.png                     | Bin 30530 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/not-available-offline-logo.png                   | Bin 811 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/offline-storage-settings.png                     | Bin 70090 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/one-folder-available-offline.png                 | Bin 33819 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/quick-access-view-nothing-available-offline.png  | Bin 76308 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/quick-access-view-toc.png                        | Bin 76591 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/regular-badges.png                               | Bin 127263 -> 0 bytes
 docs/modules/ROOT/assets/images/offline-storage/settings.png                                     | Bin 76438 -> 0 bytes
 docs/modules/ROOT/assets/images/quick-access-view.png                                            | Bin 64756 -> 42850 bytes
 docs/modules/ROOT/assets/images/sort-files-landscape-mode.png                                    | Bin 38798 -> 0 bytes
 docs/modules/ROOT/assets/images/sort-files-portrait-mode.png                                     | Bin 81529 -> 0 bytes
 docs/modules/ROOT/assets/images/themes/classic.png                                               | Bin 70728 -> 0 bytes
 docs/modules/ROOT/assets/images/themes/dark.png                                                  | Bin 92536 -> 0 bytes
 docs/modules/ROOT/assets/images/themes/light.png                                                 | Bin 63296 -> 0 bytes
 docs/modules/ROOT/nav.adoc                                                                       |   1 -
 docs/modules/ROOT/pages/ios_accounts.adoc                                                        |   7 --
 docs/modules/ROOT/pages/ios_available_offline.adoc                                               |  67 -------------
 docs/modules/ROOT/pages/ios_files.adoc                                                           |  42 +--------
 docs/modules/ROOT/pages/ios_quick_access.adoc                                                    |   3 -
 docs/modules/ROOT/pages/ios_settings.adoc                                                        |  13 +--
 docs/yarn.lock                                                                                   | 642 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++----------------------------------------------------------------
 ios-sdk                                                                                          |   2 +-
 ownCloud.xcodeproj/project.pbxproj                                                               |  22 +----
 ownCloud.xcodeproj/xcshareddata/xcschemes/ownCloud.xcscheme                                      |   5 -
 ownCloud/Client/Actions/Action.swift                                                             |   8 +-
 ownCloud/Client/Actions/EditDocumentViewController.swift                                         |  46 ++++-----
 ownCloud/Client/Actions/MoreViewHeader.swift                                                     |   5 +-
 ownCloud/Client/Actions/Scanner/ScanAction.swift                                                 |   7 +-
 ownCloud/Client/ClientItemCell.swift                                                             |   7 --
 ownCloud/Client/ClientRootViewController.swift                                                   |   4 -
 ownCloud/Client/PhotoAlbumTableViewCell.swift                                                    |   4 -
 ownCloud/Client/PhotoSelectionViewCell.swift                                                     |   4 +-
 ownCloud/Client/SortBar.swift                                                                    |   5 +-
 ownCloud/Client/Viewer/DisplayHostViewController.swift                                           |   3 +-
 ownCloud/Client/Viewer/DisplayViewController.swift                                               |  50 +---------
 ownCloud/Client/Viewer/Image/ImageDisplayViewController.swift                                    |  31 +++---
 ownCloud/Client/Viewer/Media/MediaDisplayViewController.swift                                    | 116 +++++++++++------------
 ownCloud/Client/Viewer/PDF/PDFThumbnailCollectionViewCell.swift                                  |   4 -
 ownCloud/Client/Viewer/PDF/PDFViewerViewController.swift                                         | 185 ++++++++++++++++++------------------
 ownCloud/Client/Viewer/{QuickLook => }/PreviewViewController.swift                               |  29 +++---
 ownCloud/Client/Viewer/WebView/WebViewDisplayViewController.swift                                |  49 ++++------
 ownCloud/FileLists/QueryFileListTableViewController.swift                                        |   3 +-
 ownCloud/Licensing/Offers/LicenseOfferView.swift                                                 |  16 ++--
 ownCloud/Migration/LegacyCredentials.swift                                                       |   2 +-
 ownCloud/Migration/Migration.swift                                                               |  66 +++++++++----
 ownCloud/Migration/MigrationActivityCell.swift                                                   |   2 +-
 ownCloud/Migration/MigrationViewController.swift                                                 |   2 +-
 ownCloud/PhotoKit Extensions/MediaUploadQueue.swift                                              |  51 +++++-----
 ownCloud/Release Notes/ReleaseNotes.plist                                                        |  43 ---------
 ownCloud/Resources/en-GB.lproj/Localizable.strings                                               |   2 +-
 ownCloud/Resources/en.lproj/Localizable.strings                                                  |  14 ++-
 ownCloud/SDK Extensions/OCItem+Extension.swift                                                   |  22 +----
 ownCloud/Server List/ServerListBookmarkCell.swift                                                |   4 -
 ownCloud/Settings/Passcode/PasscodeViewController.swift                                          |   8 --
 ownCloud/Theming/UI/ThemeTableViewCell.swift                                                     |  13 +--
 ownCloud/UI Elements/PointerEffect.swift                                                         |  73 --------------
 ownCloud/UI Elements/Progress/ProgressSummarizer.swift                                           |   4 +-
 ownCloud/UI Elements/StaticTableViewRow.swift                                                    |  64 ++++---------
 ownCloudAppFramework/Licensing/Offer/OCLicenseDuration.h                                         |   1 -
 ownCloudAppFramework/Licensing/Offer/OCLicenseDuration.m                                         |  62 ------------
 ownCloudAppFramework/Resources/en.lproj/Localized.strings                                        |  16 ----
 ownCloudAppShared/Intent/SaveFileIntentHandler.swift                                             |  20 ++--
 82 files changed, 673 insertions(+), 1424 deletions(-)
 delete mode 100755 docs/bin/cli
 delete mode 100644 docs/modules/ROOT/assets/images/21_File_list_parent.jpg
 delete mode 100644 docs/modules/ROOT/assets/images/file-actions.jpg
 delete mode 100644 docs/modules/ROOT/assets/images/index-bar-with-callout.png
 delete mode 100644 docs/modules/ROOT/assets/images/index-bar.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/all-available-offline-items.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/available-offline-badge-closeup-with-callout.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/available-offline-badge-closeup.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/available-offline-badge.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/available-offline-items.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/available-offline-logo.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/file-available-offline.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/folder-action.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/make-available-offline.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/make-unavailable-offline.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/not-available-offline-logo.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/offline-storage-settings.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/one-folder-available-offline.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/quick-access-view-nothing-available-offline.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/quick-access-view-toc.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/regular-badges.png
 delete mode 100644 docs/modules/ROOT/assets/images/offline-storage/settings.png
 delete mode 100644 docs/modules/ROOT/assets/images/sort-files-landscape-mode.png
 delete mode 100644 docs/modules/ROOT/assets/images/sort-files-portrait-mode.png
 delete mode 100644 docs/modules/ROOT/assets/images/themes/classic.png
 delete mode 100644 docs/modules/ROOT/assets/images/themes/dark.png
 delete mode 100644 docs/modules/ROOT/assets/images/themes/light.png
 delete mode 100644 docs/modules/ROOT/pages/ios_available_offline.adoc
 rename ownCloud/Client/Viewer/{QuickLook => }/PreviewViewController.swift (85%)
 delete mode 100644 ownCloud/UI Elements/PointerEffect.swift
```

Let me see if I can rewrite history a bit to fix that…

felix-schwarz commented:
Ok, fixed it with a `git rebase`.

Before:
<img width="1000" alt="Bildschirmfoto 2020-05-14 um 00 41 40" src="https://user-images.githubusercontent.com/639669/81873465-1f30da00-957c-11ea-81fe-d59aae9f769f.png">

After:
<img width="1001" alt="Bildschirmfoto 2020-05-14 um 00 43 40" src="https://user-images.githubusercontent.com/639669/81873472-20fa9d80-957c-11ea-9602-04611485965d.png">

Not sure if a `git pull` works to fetch the rewritten history, but please make sure your local copy is in sync with the rewritten branch before making any further changes.

felix-schwarz commented:
> @felix-schwarz thanks for feedback. Apart from bundle identifiers, I had to change all the .entitlements files manually to put the old and new app into the same app-group. Worked for me so far.

Sure enough that was it. After changing the entitlements file, everything worked fine. Thanks!

jesmrec commented:
First migration, using SAML-based server.

1. Install old app, add the account and enable passcode, face id, and instant uploads ✅ 

2. Upgrade with new version -> new app overwrites old app ✅ 

3. Migration status view in screen. Accounts and settings (instant uploads) start to migrate. It takes too long. I waited ~5 minutes and it did not finish, so i minimize the view manually... should it take longer to finish?

4. After minimizing the migration view, we have the list of bookmarks. There we have the migrated account ✅  . Settings are not migrated, everything is disabled at this point. Account is migrated, but it is empty.

5. I kill the app and reopen. Then, it shows me the migration view with "App passcode" migrated. And this is right ✅  , but i had to kill to achieve this.  I kill again to test the passcode is correct. It is ✅ 

6. I enter the bookmark again, now it asks to sign in. I sign in, account is now logged in. ✅ 

7. Instant uploads are not migrated. The migration view displayed in 3. shows that they are being migrated.

Summarizing:

- Migration is triggered, but how long does it take? is that correct?
- In first instance, nothing is migrated. Killing the app is needed to start to get the stuff migrated
- Instant uploads are not migrated
- Face ID is not migrated

I took as reference: https://github.com/owncloud/ios-app/issues/270#issuecomment-596959363

mneuwert commented:
@jesmrec Tested multiple types with same account and did only once see a delay in `OCConnection.prepareForSetup()` call probably just due to temporary server-side response delays. With latest two commits fixed two issues:

- Changed presentation style of the migration screen to avoid it's accidental dismissal by the user on iOS 13 which could have led to some unpredictable behaviour.
- Added biometrical ID migration (Face ID or Touch ID depending on the device).

Normally the migration is just a matter of running a hand full of queries on the legacy database and the only activity which can lead to delays is connecting to the server and so on. Not sure what happened in your case. Please provide logs if the issue you described is still seen.

jesmrec commented:
I was able to reproduce the problem above with the following steps

1. Create a container with docker, using image `owncloud/server:10.4.1`

`docker run -tid -p xxxx:8080 owncloud/server:10.4.1`

2. Enable OAuth2 in the container

`docker exec -ti <cont_id> occ market:install oauth2`
`docker exec -ti <cont_id> occ app:enable oauth2`

3. Install legacy app and add an account of this server
4. Install the new app to migrate

Behaviour described here is reproduced: https://github.com/owncloud/ios-app/pull/660#issuecomment-657548788

iPhoneXR v13

jesmrec commented:
Trying to scope the problem mentioned above. I was able to reproduce with the following set up:

- OAuth2 server. With basic server, it was not reproducible.

- Server URL with issues to fix by the user (certificate, f.ex). If server does not have issues (like `demo.owncloud.com`), not reproducible

- Any of the instant uploads settings is enabled. If instant uploads are disabled, not reproducible

If server matches with these three, i reproduce the problem. If any of them does not match, everything works fine.

I hope it will help. If not, we can go for the other side: enabling logging by default, so i can gather logs from scratch.

jesmrec commented:
These are the logs i caught. I hope it helps:

[ownCloud_4_Aug_2020_at_18_20_59.log.txt](https://github.com/owncloud/ios-app/files/5023271/ownCloud_4_Aug_2020_at_18_20_59.log.txt)

Steps:

1. In legacy app, add an OAuth2 app which URL has issues to approve (self signed cert, f. ex). Steps to create [here](https://github.com/owncloud/ios-app/pull/660#issuecomment-659242684)
2. Enable `Instant Uploads Photos` (no more settings enabled)
3. Install the new app (current branch) to upgrade

Current: URL ✅  ,  instant uploads ⚙️  forever

I used XCode11.6 to install both apps. Legacy app with branch `migration_test` and new app with current branch. 

iPhoneXR v13.6

mneuwert commented:
@jesmrec thanks for checking... will try to reproduce

mneuwert commented:
@jesmrec Found and fixed issue leading to "Instant upload settings" migration hanging forever

jesmrec commented:
Problem is finally fixed
